doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+
+### <a name="design-components-dimension-tags"></a>Dimension tags
+
+A dimension is identified by a tag.
+This is simply an `enum class Dim`.
+*Note:
+In the prototype this is currently named `enum class Dimension`, but we suggest renaming this to `enum class Dim` for brevity.
+Despite the abbreviation this should be clear and may actually improve code readability since it is more concise when used frequently?*
+
+An incomplete list of dimension tags:
+- `Dim::X`, `Dim::Y`, `Dim::Z`.
+- `Dim::Qx`, `Dim::Qy`, `Dim::Qz`.
+- `Dim::Detector`
+- `Dim::Spectrum`
+- `Dim::Tof`
+- `Dim::Wavelength`
+- `Dim::DetectorScan`
+- `Dim::Temperature`
+- `Dim::Polarization`
+- `Dim::Run`
+- `Dim::Row`
+- `Dim::Event`
+
+In many cases there is a corresponding coordinate tag for a dimension.
+
+
+### <a name="design-components-datasetview"></a>`class DatasetView`
+
+Due to the type-erasure we cannot simply treat `Dataset` as an array of structs, i.e., there is no equivalent to `API::MatrixWorkspace::getSpectrum(const size_t index)`.
+If only access to a single variable is required it can be accessed directly.
+For cases that require joint access to multiple variables we provide `class DatasetView`:
+- Effectively provides a "zip-iterator", with support for automatic on-the-fly broadcast or transposition of dimensions.
+- Nested view allows for accessing a histogram-like view.
+- Allows for joint iteration with bin-edge variables using a `Bin` wrapper, overcoming the old nuisance of having `length+1` bin edges.
+  See the [C++ example](#examples-cpp) for an example of how this would be used.
+
+The current implementation seems to support all required features.
+Performance was originally a problem due to the multi-dimensional iteration, but has subsequently been optimized to the point where it should be sufficient for most applications.
+
+
+### <a name="design-components-datasetindex"></a>`class DatasetIndex`
+
+`class DatasetIndex` is a simple helper to index slices in a dataset via values in an axis (coordinate).
+`Dataset` itself makes no attempt at keeping values in a coordinate variable validated.
+For example, there is no check for duplicate spectrum numbers.
+Therefore, direct indexing is not possible.
+Instead `DatasetIndex` builds the index (and does the validation) on demand when constructed.
+
+##### Rationale
+
+I tried to *maintain* consistent indices like spectrum numbers using code in `Indexing::IndexInfo`, but it turned out to be really complicated when integrating it into `API::MatrixWorkspace`.
+
+
+### <a name="design-components-units"></a>Units
+
+Internally the prototype uses `boost::units`, but we require run-time units.
+It is therefore wrapped in a class that simply stores an unique ID for every unit.
+- Arithmetic operations branch off using the ID, relying on `boost::units` for the actual work.
+- Probably need to add support for units scales.
+
+Very little of this is implemented in the prototype apart from the most basic examples.
```

Comment:
How would unit conversion work, for example `TOF`->`dSpacing`?

Reply:
As discussed, the unit is linked to the dimension. Let us consider your example: `Tof` -> `dSpacing`:

- We start with a dataset with `Dim::Tof` as one of its dimensions.
- There is a coordinate for this dimension, `Coord::Tof`.
  - Has `Dim::Tof`, but there can be others.
  - Has `Unit::Tof`.

The implementation could look something like this:

1. Convert the coordinate:
   ```cpp
   dataset.insert<Coord::dSpacing>(convert(dataset.erase<Coord::Tof>(), Dim::dSpacing));
   ```
2. Rename the dimension of all other variables in the dataset:
   ```cpp
   for (auto &var : dataset)
     var.dimensions().rename(Dim::Tof, Dim::dSpacing);
   ```

(this is assuming the variables are not "distributions", which would also require converting the data variables).